### DISCUSSION TITLE
"received NO_PROPOSAL_CHOSEN notify" on FreeBSD

### DISCUSSION THREAD

**Discussion by sonyarpita** [strongswan/strongswan#339] [Q&A]
Following is the configuration on Host1 - 
root@aqua3:/usr/local/etc # cat ipsec.conf
config setup
        strictcrlpolicy=no

conn %default
 ikelifetime=60m
 keylife=20m
 aggressive=yes
 rekeymargin=3m
 keyingtries=1
 mobike=no
 keyexchange=ikev2
 ike=aes-sha1-modp1024!
 esp=aes256gcm16!

conn aqua3-aqua5-p0
    left=102.1.1.189
    leftcert=aqua3Cert.der
    leftid="C=CH, O=blr.asicdesigners.com, CN=aqua3"
    leftfirewall=yes
    right=102.1.1.87
    rightid="C=CH, O=blr.asicdesigners.com, CN=aqua5"
    type=transport
    auto=add

Configuration on Host2 - 
root@aqua5:/usr/local/etc # cat ipsec.conf
config setup
        strictcrlpolicy=no

conn %default
 ikelifetime=60m
 keylife=20m
 aggressive=yes
 rekeymargin=3m
 keyingtries=1
 mobike=no
 keyexchange=ikev2
 ike=aes-sha1-modp1024!
 esp=aes256gcm16!

conn aqua3-aqua5-p0
    left=102.1.1.87
    leftcert=aqua5Cert.der
    leftid="C=CH, O=blr.asicdesigners.com, CN=aqua5"
    leftfirewall=yes
    right=102.1.1.189
    rightid="C=CH, O=blr.asicdesigners.com, CN=aqua3"
    type=tunnel
    auto=add

Following error is observed on trying to establish IPSec tunnel - [However the above configuration works on between 2 Linux machines ] 
root@aqua3:/usr/local/etc # ipsec start
Starting strongSwan 5.9.2 IPsec [starter]...

root@aqua3:/usr/local/etc # ipsec up aqua3-aqua5-p0
initiating IKE_SA aqua3-aqua5-p0[1] to 102.1.1.87
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 102.1.1.189[500] to 102.1.1.87[500] (336 bytes)
received packet: from 102.1.1.87[500] to 102.1.1.189[500] (389 bytes)
parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
selected proposal: IKE:AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
received cert request for "C=CH, O=blr.asicdesigners.com, CN=blr.asicdesigners.com CA"
received cert request for "C=CH, O=blr.asicdesigners.com, CN=blr.asicdesigners.com CAp1"
sending cert request for "C=CH, O=blr.asicdesigners.com, CN=blr.asicdesigners.com CA"
sending cert request for "C=CH, O=blr.asicdesigners.com, CN=blr.asicdesigners.com CAp1"
authentication of 'C=CH, O=blr.asicdesigners.com, CN=aqua3' (myself) with RSA_EMSA_PKCS1_SHA2_256 successful
sending end entity cert "C=CH, O=blr.asicdesigners.com, CN=aqua3"
establishing CHILD_SA aqua3-aqua5-p0{1}
generating IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) CERTREQ IDr AUTH N(USE_TRANSP) N(ESP_TFC_PAD_N) SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
splitting IKE message (1500 bytes) into 2 fragments
generating IKE_AUTH request 1 [ EF(1/2) ]
generating IKE_AUTH request 1 [ EF(2/2) ]
sending packet: from 102.1.1.189[500] to 102.1.1.87[500] (1248 bytes)
sending packet: from 102.1.1.189[500] to 102.1.1.87[500] (320 bytes)
received packet: from 102.1.1.87[500] to 102.1.1.189[500] (1248 bytes)
parsed IKE_AUTH response 1 [ EF(1/2) ]
received fragment #1 of 2, waiting for complete IKE message
received packet: from 102.1.1.87[500] to 102.1.1.189[500] (80 bytes)
parsed IKE_AUTH response 1 [ EF(2/2) ]
received fragment #2 of 2, reassembled fragmented IKE message (1260 bytes)
parsed IKE_AUTH response 1 [ IDr CERT AUTH N(AUTH_LFT) N(NO_PROP) ]
received end entity cert "C=CH, O=blr.asicdesigners.com, CN=aqua5"
  using certificate "C=CH, O=blr.asicdesigners.com, CN=aqua5"
  using trusted ca certificate "C=CH, O=blr.asicdesigners.com, CN=blr.asicdesigners.com CA"
checking certificate status of "C=CH, O=blr.asicdesigners.com, CN=aqua5"
certificate status is not available
  reached self-signed root ca with a path length of 0
authentication of 'C=CH, O=blr.asicdesigners.com, CN=aqua5' with RSA_EMSA_PKCS1_SHA2_256 successful
IKE_SA aqua3-aqua5-p0[1] established between 102.1.1.189[C=CH, O=blr.asicdesigners.com, CN=aqua3]...102.1.1.87[C=CH, O=blr.asicdesigners.com, CN=aqua5]
scheduling reauthentication in 3353s
maximum IKE_SA lifetime 3533s
received NO_PROPOSAL_CHOSEN notify, no CHILD_SA built
failed to establish CHILD_SA, keeping IKE_SA
establishing connection 'aqua3-aqua5-p0' failed

root@aqua3:/usr/local/etc # ipsec version
FreeBSD strongSwan U5.9.2/K14.0-CURRENT
University of Applied Sciences Rapperswil, Switzerland
See 'ipsec --copyright' for copyright information.

**Comment by Thermi** [comment]
> received NO_PROPOSAL_CHOSEN notify, no CHILD_SA built

Issue is on the remote peer. Check logs there.

**Comment by sonyarpita** [comment]
[charon.log](https://github.com/strongswan/strongswan/files/6347758/charon.log)
Here is the log from the Peer, I could not find anything suspicious . 

And also, even though connection fails to establish with ipsec up <connection_name> , ipsec statusall shows connection as "ESTABLISHED" 

root@aqua3:~ # ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.2, FreeBSD 14.0-CURRENT, amd64):
  uptime: 26 minutes, since Apr 21 08:15:55 2021
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 3
  loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem fips-prf gmp curve25519 xcbc cmac hmac drbg attr kernel-pfkey kernel-pfroute resolve socket-default stroke vici updown xauth-generic counters
Listening IP addresses:
  10.193.80.189
  102.1.1.189
  1000::189
  102.2.2.189
  2000::189
Connections:
aqua3-aqua5-p0:  102.1.1.189...102.1.1.87  IKEv2
aqua3-aqua5-p0:   local:  [C=CH, O=blr.asicdesigners.com, CN=aqua3] uses public key authentication
aqua3-aqua5-p0:    cert:  "C=CH, O=blr.asicdesigners.com, CN=aqua3"
aqua3-aqua5-p0:   remote: [C=CH, O=blr.asicdesigners.com, CN=aqua5] uses public key authentication
aqua3-aqua5-p0:   child:  dynamic === dynamic TRANSPORT
Security Associations (1 up, 0 connecting):
aqua3-aqua5-p0[1]: ESTABLISHED 26 minutes ago, 102.1.1.189[C=CH, O=blr.asicdesigners.com, CN=aqua3]...102.1.1.87[C=CH, O=blr.asicdesigners.com, CN=aqua5]
aqua3-aqua5-p0[1]: IKEv2 SPIs: 8c97c498ff446dac_i* 56863c6f8f3d8257_r, public key reauthentication in 26 minutes
aqua3-aqua5-p0[1]: IKE proposal: AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024

**Reply by Thermi** [reply]
> And also, even though connection fails to establish with ipsec up <connection_name> , ipsec statusall shows connection as "ESTABLISHED"

The IPsec architecture consists of an IKE SA for management and an arbitrary number CHILD SAs that are used to process the traffic. The ESTABLISHED here is for the IKE SA.

Also, why do you try to configure technology you do not understand the basics of in a scenario that you obviously do not have guidance for?

> Here is the log from the Peer, I could not find anything suspicious .

Please. There's something wrong with your kernel. Probably a missing config setting at build time.

```
Apr 21 08:16:03 01[JOB] watched FD 9 ready to read
Apr 21 08:16:03 05[KNL] <aqua3-aqua5-p0|1> adding SAD entry with SPI c9a04a8c and reqid {1}
Apr 21 08:16:03 01[JOB] watcher going to poll() 4 fds
Apr 21 08:16:03 05[KNL] <aqua3-aqua5-p0|1>   using encryption algorithm AES_GCM_16 with key size 288
Apr 21 08:16:03 05[KNL] <aqua3-aqua5-p0|1> unable to add SAD entry with SPI c9a04a8c: Invalid argument (22)
Apr 21 08:16:03 01[JOB] watcher got notification, rebuilding
Apr 21 08:16:03 05[CHD] <aqua3-aqua5-p0|1> adding outbound ESP SA
Apr 21 08:16:03 01[JOB] watcher going to poll() 5 fds
Apr 21 08:16:03 05[CHD] <aqua3-aqua5-p0|1>   SPI 0xc28046db, src 102.1.1.87 dst 102.1.1.189
Apr 21 08:16:03 05[KNL] <aqua3-aqua5-p0|1> adding SAD entry with SPI c28046db and reqid {1}
Apr 21 08:16:03 05[KNL] <aqua3-aqua5-p0|1>   using encryption algorithm AES_GCM_16 with key size 288
Apr 21 08:16:03 05[KNL] <aqua3-aqua5-p0|1> unable to add SAD entry with SPI c28046db: Invalid argument (22)
Apr 21 08:16:03 05[IKE] <aqua3-aqua5-p0|1> unable to install inbound and outbound IPsec SA (SAD) in kernel
Apr 21 08:16:03 05[ENC] <aqua3-aqua5-p0|1> added payload of type NOTIFY to message
Apr 21 08:16:03 05[IKE] <aqua3-aqua5-p0|1> failed to establish CHILD_SA, keeping IKE_SA
Apr 21 08:16:03 05[CHD] <aqua3-aqua5-p0|1> CHILD_SA aqua3-aqua5-p0{1} state change: INSTALLING => DESTROYING
```

**Reply by sonyarpita** [reply]
> The IPsec architecture consists of an IKE SA for management and an arbitrary number CHILD SAs that are used to process the traffic. The ESTABLISHED here is for the IKE SA.
> 
> Also, why do you try to configure technology you do not understand the basics of in a scenario that you obviously do not have guidance for?
> 
I agree with your inference and I will try and understand this technology and really appreciate the nudge 

Thank you for all the patience.
